velox/substrait/TypeUtils.cpp

```diff
@@ -17,45 +17,87 @@
+namespace {
+std::vector<std::string> getRowTypesFromCompoundName(
+    const std::string& compoundName) {
+  // CompoundName is like ROW<BIGINT,DOUBLE>
+  // or ROW<BIGINT,ROW<DOUBLE,BIGINT>,ROW<DOUBLE,BIGINT>>
+  // the position of then delimiter is where the number of leftAngleBracket
+  // equals rightAngleBracket need to split.
+  std::vector<std::string> types;
+  std::vector<int> angleBracketNumEqualPos;
+  auto leftAngleBracketPos = compoundName.find("<");
+  auto rightAngleBracketPos = compoundName.rfind(">");
+  auto typesName = compoundName.substr(
+      leftAngleBracketPos + 1, rightAngleBracketPos - leftAngleBracketPos - 1);
+  int leftAngleBracketNum = 0;
+  int rightAngleBracketNum = 0;
+  for (auto index = 0; index < typesName.length(); index++) {
+    if (typesName[index] == '<') {
+      leftAngleBracketNum++;
+    }
+    if (typesName[index] == '>') {
+      rightAngleBracketNum++;
+    }
+    if (typesName[index] == ',' &&
+        rightAngleBracketNum == leftAngleBracketNum) {
+      angleBracketNumEqualPos.push_back(index);
+    }
+  }
+  int startPos = 0;
+  for (auto delimeterPos : angleBracketNumEqualPos) {
+    types.emplace_back(typesName.substr(startPos, delimeterPos - startPos));
+    startPos = delimeterPos + 1;
   }
-  return false;
+  types.emplace_back(typesName.substr(startPos, typesName.length() - startPos));
+  return types;
 }
 
-int64_t bytesOfType(const TypePtr& type) {
-  auto typeKind = type->kind();
-  switch (typeKind) {
-    case TypeKind::INTEGER:
-      return 4;
-    case TypeKind::BIGINT:
-    case TypeKind::REAL:
-    case TypeKind::DOUBLE:
-      return 8;
-    default:
-      VELOX_NYI("Returning bytes of Type not supported for type {}.", typeKind);
+std::string getNameBeforeDelimiter(
+    const std::string& compoundName,
+    const std::string& delimiter) {
+  std::size_t pos = compoundName.find(delimiter);
+  if (pos == std::string::npos) {
+    return compoundName;
   }
+  return compoundName.substr(0, pos);
 }
+} // namespace
 
 TypePtr toVeloxType(const std::string& typeName) {
-  auto typeKind = mapNameToTypeKind(typeName);
+  auto type = getNameBeforeDelimiter(typeName, "<");
+  auto typeKind = mapNameToTypeKind(type);
   switch (typeKind) {
     case TypeKind::BOOLEAN:
       return BOOLEAN();
-    case TypeKind::DOUBLE:
-      return DOUBLE();
-    case TypeKind::VARCHAR:
-      return VARCHAR();
+    case TypeKind::TINYINT:
+      return TINYINT();
+    case TypeKind::SMALLINT:
+      return SMALLINT();
     case TypeKind::INTEGER:
       return INTEGER();
     case TypeKind::BIGINT:
       return BIGINT();
+    case TypeKind::REAL:
+      return REAL();
+    case TypeKind::DOUBLE:
+      return DOUBLE();
+    case TypeKind::VARCHAR:
+      return VARCHAR();
+    case TypeKind::VARBINARY:
+      return VARBINARY();
+    case TypeKind::ROW: {
+      auto fieldNames = getRowTypesFromCompoundName(typeName);
+      VELOX_CHECK(!fieldNames.empty(), "At lease one type name is expected.");
```

Comment:
giving more context is always useful "... expected when converting Substrait to Velox type."